src/types.ts

```diff
@@ -288,8 +288,6 @@ export namespace Server {
   export interface LobbyConfig {
     uuid?: () => string;
     generateCredentials?: GenerateCredentials;
-    apiPort?: number;
-    apiCallback?: () => void;
```

Comment:
Why are `apiPort` and `apiCallback` removed here? They’re still in use as far as I can tell.

Reply:
The only place where these values seem to be used is in `Server.run` and `createServerRunConfig` (which makes sense since `apiPort` and `apiCallback` are very much runtime considerations as opposed to generic configurations). However, `createServerRunConfig` has its own type definition for `apiPort` and `apiCallback` in the `lobbyConfig` member of `ServerConfig`. This means that the values could be safely removed from the more generic `LobbyConfig` interface.

To put this in another way: previously `LobbyConfig` was really two distinct interfaces that we could have called `LobbySettings` (`uuid` and `generateCredentials`, used by the framework code) and `LobbyRuntimeSettings` (`apiPort` and `apiCallback`, to my understanding only used by `Server.run`). This pull request encapsulates the `LobbyRuntimeSettings` in `Server.run`.

Reply:
Ahh, got it! Thanks. Could we type the `portOrConfig` argument to `run` on line 119 to reflect this? (It’s currently just `number | object`.)

Reply:
Oh, and we should probably update the documentation with the `lobbyConfig` option to `Server`.

Reply:
Done.

Reply:
Note that there are now quite a few redundant arguments in the Server function arguments:

- `https` and `authenticateCredentials` are less specific than a custom `transport`.
- `generateCredentials` is less specific than `lobbyConfig`.

If you're open to breaking API changes, I would recommend to deprecate `https`, `authenticateCredentials` and `generateCredentials` in favor of `transport` and `lobbyConfig` to reduce future maintenance burden and confusion with the function arguments (e.g. which argument takes precedence). If you'd like, I can make a PR for these changes.

Reply:
Thanks! Aren’t these already breaking changes as `uuid` is moved from the `lobbyConfig` passed to `server.run` to the `Server` constructor options?

I’ve done some more refactoring here to reduce some of the specificity issues you raise: https://github.com/nicolodavis/boardgame.io/tree/delucis/pr/698

If you think that looks good, I’ll add it to this PR and get this merged.

In terms of further changes, do you have any views on #693?

Reply:
While I believe it was technically possible to pass `uuid` and `generateCredentials` via the `Server.run` argument, this wasn't expressed in the type system. I'll make a change to make this fully backwards compatible.

Reply:
Typescript is a fairly recent addition to the library, so there are a few places where the types don’t fully reflect potential usage out there.

That said, we’re actually planning to release several breaking changes in 0.40.0 so I think this is a reasonable time to clean this up as long as we document the migration path clearly. I’m going to merge your earlier version into a new development branch from which we can then release this.

Thanks again!